src/storage/heap_file.c

```diff
@@ -21136,45 +21218,105 @@ heap_delete_home (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context, boo
       char data_buffer[IO_DEFAULT_PAGE_SIZE + OR_MVCC_MAX_HEADER_SIZE + MAX_ALIGNMENT];
       int header_size, adjusted_size;
       bool is_adjusted_size_big = false;
+      int offset, repid_and_flag_bits, mvcc_flags;
+      char *build_recdes_data;
+      bool use_optimization;
+      int sizeof_log_lsa = sizeof (LOG_LSA);
+
+      /* Build the new record descriptor. */
+      repid_and_flag_bits = OR_GET_MVCC_REPID_AND_FLAG (context->home_recdes.data);
+      mvcc_flags = (repid_and_flag_bits >> OR_MVCC_FLAG_SHIFT_BITS) & OR_MVCC_FLAG_MASK;
+      header_size = or_mvcc_header_size_from_flags (mvcc_flags);
+      adjusted_size = context->home_recdes.length;
 
-      /* get home record header size */
-      error_code = or_mvcc_get_header (&context->home_recdes, &record_header);
-      if (error_code != NO_ERROR)
+      /* Uses the optimization in most common cases, for now : if DELID not set and adjusted size is not big size. */
+      use_optimization = true;
+      if (!(mvcc_flags & OR_MVCC_FLAG_VALID_DELID))
 	{
-	  ASSERT_ERROR ();
-	  return error_code;
+	  adjusted_size += OR_MVCCID_SIZE;
+	  is_adjusted_size_big = heap_is_big_length (adjusted_size);
+	  if (is_adjusted_size_big)
+	    {
+	      /* Rare case, do not optimize it now. */
+	      use_optimization = false;
+	    }
 	}
-      header_size = or_mvcc_header_size_from_flags (record_header.mvcc_flag);
-      adjusted_size = context->home_recdes.length;
-      if (!MVCC_IS_FLAG_SET (&record_header, OR_MVCC_FLAG_VALID_DELID))
+      else
 	{
-	  adjusted_size += OR_MVCCID_SIZE;	/* delid size */
+	  /* Rare case, do not optimize it now. */
+	  is_adjusted_size_big = false;
+	  use_optimization = false;
 	}
 
-      is_adjusted_size_big = heap_is_big_length (adjusted_size);
-#if !defined(NDEBUG)
-      if (is_adjusted_size_big)
+      /* Build the new record. */
+      HEAP_SET_RECORD (&built_recdes, IO_DEFAULT_PAGE_SIZE + OR_MVCC_MAX_HEADER_SIZE, 0, REC_UNKNOWN,
+		       PTR_ALIGN (data_buffer, MAX_ALIGNMENT));
+      if (use_optimization)
```

Comment:
Please examine whether we can rewrite `if (use_optimization)` block as follows:

```C
      if (use_optimization)
        {
          char *start_p;

          delid_offset = OR_MVCC_DELETE_ID_OFFSET (mvcc_flags);

          build_recdes_data = start_p = built_recdes.data;

          /* Copy up to MVCC DELID first. */
          memcpy (build_recdes_data, context->home_recdes.data, delid_offset);
          build_recdes_data += delid_offset;

          /* Sets MVCC DELID flag, overwrite first four bytes. */
          repid_and_flag_bits |= (OR_MVCC_FLAG_VALID_DELID << OR_MVCC_FLAG_SHIFT_BITS);
          OR_PUT_INT (start_p, repid_and_flag_bits);

          /* Sets the MVCC DELID. */
          OR_PUT_BIGINT (build_recdes_data, &mvcc_id);
          build_recdes_data += OR_MVCC_DELETE_ID_SIZE;

          /* Sets the previous version if exists. */
          if (mvcc_flags & OR_MVCC_FLAG_VALID_PREV_VERSION)
            {
              assert (delid_offset == OR_MVCC_PREV_VERSION_LSA_OFFSET (mvcc_flags));
              memcpy (build_recdes_data, context->home_recdes.data + delid_offset, OR_MVCC_PREV_VERSION_LSA_SIZE);
            }

          /* Copy behind header. */
          memcpy (build_recdes_data, context->home_recdes.data + header_size,
                  context->home_recdes.length - header_size);
          built_recdes.length = adjusted_size;
        }
```

Please also try to rewrite the `if` block of `heap_delete_relocation`.